Sources/BrowserServicesKit/LinkProtection/LinkCleaner.swift

```diff
@@ -0,0 +1,119 @@
+                return format
+            }
+        }
+        
+        return nil
+    }
+    
+    public func isURLExcluded(url: URL, feature: PrivacyFeature = .ampLinks) -> Bool {
+        guard let host = url.host else { return true }
+        
+        return !privacyConfig.isFeature(feature, enabledForDomain: host)
+    }
+    
+    public func extractCanonicalFromAMPLink(initiator: URL?, destination url: URL?) -> URL? {
+        lastAMPURLString = nil
+        guard privacyConfig.isEnabled(featureKey: .ampLinks) else { return nil }
+        guard let url = url, !isURLExcluded(url: url) else { return nil }
+        if let initiator = initiator, isURLExcluded(url: initiator) {
+            return nil
+        }
+        
+        guard let ampFormat = ampFormat(matching: url) else { return nil }
+        
+        do {
+            let ampStr = url.absoluteString
+            let regex = try NSRegularExpression(pattern: ampFormat, options: [.caseInsensitive])
+            let matches = regex.matches(in: url.absoluteString,
+                                        options: [],
+                                        range: NSRange(ampStr.startIndex..<ampStr.endIndex,
+                                                       in: ampStr))
+            guard let match = matches.first else { return nil }
+            
+            let matchRange = match.range(at: 1)
+            if let substrRange = Range(matchRange, in: ampStr) {
+                var urlStr = String(ampStr[substrRange])
+                if !urlStr.hasPrefix("http") {
+                    urlStr = "https://\(urlStr)"
+                }
+                
+                if let cleanUrl = URL(string: urlStr), !isURLExcluded(url: cleanUrl) {
+                    lastAMPURLString = ampStr
+                    return cleanUrl
+                }
+            }
+        } catch {
+            return nil
+        }
+        
+        return nil
+    }
+    
+    public func cleanTrackingParameters(initiator: URL?, url: URL?) -> URL? {
+        urlParametersRemoved = false
+        guard privacyConfig.isEnabled(featureKey: .trackingParameters) else { return url }
+        guard let url = url, !isURLExcluded(url: url, feature: .trackingParameters) else { return url }
+        if let initiator = initiator, isURLExcluded(url: initiator, feature: .trackingParameters) {
+            return url
+        }
+        
+        guard var urlsComps = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+            return url
+        }
+        guard let queryParams = urlsComps.queryItems, queryParams.count > 0 else {
+            return url
+        }
+        
+        let trackingParams = TrackingLinkSettings(fromConfig: privacyConfig).trackingParameters
+        
+        let preservedParams: [URLQueryItem] = queryParams.filter { param in
+            if trackingParams.contains(where: { param.name.matches(pattern: "^\($0)$") }) {
```

Comment:
Please check out this PR on the iOS repo: https://github.com/duckduckgo/iOS/pull/1096

I think I updated this while you were working on this project. These regexes need to be case sensitive.